### ISSUE TITLE
Check if VT-X is enabled and improve related documentation

### ISSUE THREAD

**Issue by thaJeztah** [docker/toolbox#166]
Docker Toolbox (VirtualBox) fails to run if VT-X isn't supported or not enabled on the hardware it runs on. This is a placeholder issue for improving the checks, feedback and instructions to handle hardware that doesn't have VT-X (enabled).

Docker Toolbox should;

- Check if VT-X is enabled (for OS X, this article http://kb.parallels.com/en/5653 explains checking via `sysctl -a | grep machdep.cpu.features`)
- (better) Document that VT-X should be supported *and enabled*
- Add instructions / documentation how to check if VT-X is enabled and/or enable it (for OS X; https://support.apple.com/en-us/HT203296)

Given that there's an overlap between "Docker Machine", "Docker Toolbox", and "Boot2Docker", documenting it in the right places could be a challenge (users are reporting these issues in all repos :smile:)

VT-X related issues are already mentioned in https://github.com/docker/toolbox/issues/65, https://github.com/docker/toolbox/issues/119, and reported in the docker repository; https://github.com/docker/docker/issues/11877, https://github.com/docker/docker/issues/15712, https://github.com/docker/docker/issues/15982

**Comment by mchiang0610** [comment]
@thaJeztah Thanks for submitting this. This is something we definitely need.

**Comment by thaJeztah** [comment]
@mchiang0610 this could be useful information as well; https://github.com/docker/docker/issues/16182#issuecomment-138958790

**Comment by mchiang0610** [comment]
@thaJeztah Thanks. New Macs have VT-X enabled by default. We haven't seen any problems there. 

Many Windows users on the other hand have experienced this issue. Some workstations like Microsoft's Surface Pro 3 has VT-X enabled by default. 

I have added a readme change to help users. See PR here: https://github.com/docker/toolbox/pull/170

**Comment by thaJeztah** [comment]
You may be right on the OS X part, I think all that's needed is to have the firmware up to date (I see https://github.com/docker/docker/issues/11877#issuecomment-94556037 is running OS X on Hackintosh).

Still, if the installer would be able to detect it, that'd be nice :-)

**Comment by mchiang0610** [comment]
@thaJeztah Agreed! That will come. It's just that at the moment we are focused on bug fixes to cut down failure rates. :)

**Comment by thaJeztah** [comment]
:+1: sgtm

**Comment by dmp42** [comment]
cc @dgageot @jeanlaurent

**Comment by SvenDowideit** [comment]
It'd be nice if there was an explanation if I've accidentally started my windows box with hyper-v on (as I have to choose the other windows boot option) - or even better, that the quickstart-cli just picks a hyper-v based vm

**Comment by aro-tech** [comment]
Since a Windows update in windows 10 this morning, the docker quickstart falsely detects that  virtualization is disabled. My windows task manager says virtualization is active, but in the quickstart I get:

Starting "default"...
This computer doesn't have VT-X/AMD-v enabled. Enabling it in the BIOS is mandatory
Looks like something went wrong... Press any key to continue...

**Comment by dgageot** [comment]
Hi @aro-tech which version of Toolbox/docker machine are you using?

**Comment by dgageot** [comment]
@aro-tech Could you copy to a [gist](https://gist.github.com/) the output of `docker-machine create -d virtualbox test`?

**Comment by aro-tech** [comment]
[Gist, including error output and docker-machine version](https://gist.github.com/aro-tech/b21baa9c195acedb3cf9)

**Comment by aro-tech** [comment]
@dgageot The toolbox version is 1.9.1i - installed without issue in late January.  As I said it stopped working a few days ago after a Windows update. I will try installing 1.10.1.

**Comment by aro-tech** [comment]
[Same issue on version 1.10.1](https://gist.github.com/aro-tech/44fcb37db0425f74e421)

**Comment by dgageot** [comment]
@aro-tech Can you try this command? `docker-machine create -d virtualbox --virtualbox-no-vtx-check test`

**Comment by dgageot** [comment]
@aro-tech Could you also print the output of `wmic cpu get VirtualizationFirmwareEnabled`?

**Comment by aro-tech** [comment]
@dgageot [Output in a new gist](https://gist.github.com/aro-tech/f95d9ade6c9c06d70b60).

**Comment by dgageot** [comment]
Thank you very much for this feedback @aro-tech. Looks like the pre-create checks are ok. So it's not that docker-machine thinks VT-x is disabled. It's vbox that fails to start the VM. So that's not good news.

Can you gist the Vbox.log that is in `~/.docker/machine/machines/default/default/Logs/VBox.log`? It should contain the vbox error message.

Also, for completeness, can you run `docker-machine -D create -d virtualbox test`?

**Comment by aro-tech** [comment]
Here's [the log](https://gist.github.com/aro-tech/c5e458f3083da0ed9460). 
And here's [the debug output](https://gist.github.com/aro-tech/68d9a920d1c7571f0353).

**Comment by dgageot** [comment]
Thank you @aro-tech. The line that triggers the error is `... Falling back to raw-mode: VT-x is disabled in the BIOS for all CPU modes`

The virtualbox driver code thinks that `VT-x is disabled` mean the VM won't boot. In fact it might very well boot but in raw (slower) mode.

I'm going to improve the error detection. In the mean time, there's something on your computer that changed and makes vbox think VT-x is not enabled and you should try to find out what it is. I have no clue for now.

**Comment by dgageot** [comment]
Will close the issue here since I created a ticket on `docker-machine`: https://github.com/docker/machine/issues/3065

**Comment by aro-tech** [comment]
@dgageot Thanks for your assistance. At least I've learned a few things.

Personal aside: By coincidence, the README for [my little open source github project](https://github.com/aro-tech/interface-it) acknowledges an article on your blog as an inspiration for the project.

**Comment by dgageot** [comment]
\o/

**Comment by aro-tech** [comment]
Happy ending: The "It Crowd" solution has fixed the issue. After

**Comment by thaJeztah** [comment]
![](https://fangirlfatale.files.wordpress.com/2014/09/it-crowd-2.jpg)

**Comment by dgageot** [comment]
Please reboot the Internet then

**Comment by lattwood** [comment]
We've run into an issue with this on Apple equipment using Boot Camp- you have to boot into OS X (OS X writes to the MSR to enable VT-X), then warm reboot into Windows.

I might be opening a pandora's box here, but why is hardware accelerated virtualization required for Docker Toolbox? Virtualbox has support for software.

**Comment by kevinBlum** [comment]
Hi, I was looking at this issue for need of documentation and as a beginner was hoping to contribute to the project, I noticed the issue under docker/docker issue #16182 and was wondering if this issue still needed proper documentation because I would be happy to write it up

**Comment by mcandre** [comment]
No matter which options I select, Kitematic complains:

> This computer doesn't have VT-X/AMD-v enabled. Enabling it in the BIOS is mandatory

* Docker Toolbox 1.11.2
* Boot Camp 6.0
* Windows 10 Professional x86_64
* 13" mid-2013 MacBook Air, Intel Haswell Core i7

Kitematic works fine when I reboot into Mac OS El Capitan, so I'm curious if the Boot Camp drivers fail to pass through this capability?

**Comment by lattwood** [comment]
@mcandre you're not going to like the solution

Boot into OS X *once*. Then soft reboot into Windows.

PCs enable the VT-x processor flag before handing control over to the bootloader.

Macs leave it up to the operating system.

In Windows, you need kernel level access in order to toggle that processor flag, so you're SOL in terms of getting it working in a direct boot to Windows situation.

**Comment by mcandre** [comment]
@lattwood Thanks for the detailed workaround. I wish Boot Camp setup a minimal bootloader to fix these kinds of issues.

**Comment by fxd1111359** [comment]
@mchiang0610 Hi, I just found the VT-x issue on my Surface Pro 4 and couldn't find a good solution online. When I run the Docker quickstart terminal, it shows:
    **Running pre-create checks...
    Error with pre-create check: "This computer doesn't have VT-X/AMD-v enabled. Enabling it in the BIOS is mandatory"
    Looks like something went wrong in step ´Checking if machine default exists´... Press any key to continue...**
But on my Surface Pro 4, there is no option on BIOS/UEFI which can let you enable VT-x. Meanwhile, I found in Task Manager->Performance, the "Virtualization" is "Enabled". I saw your [link ](http://www.howtogeek.com/213795/how-to-enable-intel-vt-x-in-your-computers-bios-or-uefi-firmware)in PR 170, but nothing can help. Any idea on this? Thanks

**Comment by alvarezskinner** [comment]
I have the same issue in Windows 10 with my surface book, just for curiosity I installed Ubuntu on it and it worked fine. I believe this is something in the latest win10 updates that is screwing up virtual box and the vt-x detection. Other software will keep complaining hyper-V is present even when I have the feature disabled.

**Comment by GizmoAaron** [comment]
Hi, I've also ran into this issue of VT-X not enabled, while my Task Manager will show that Visualization is enabled. Using QuickStart Terminal I got nowhere, however I ran 
docker-machine create -d virtualbox --virtualbox-no-vtx-check test
and everything seemed to work fine, docker prompted me to download things, and my "Docker is up and running!", as it tells me. So is this a valid way to start my docker machine or will this eventually lead to problems down the line?
SideNote: This is my first time using Docker.

**Comment by Ronkiro** [comment]
Also had this problem.

If you have virtualization working, i see no problems @GizmoAaron
